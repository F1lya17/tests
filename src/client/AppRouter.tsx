import React from "react";
import { Route, Switch } from "react-router";
import { Home } from "./pages/Home";
import { Catalog } from "./pages/Catalog";
import { Product } from "./pages/Product";
import { Delivery } from "./pages/Delivery";
import { Contacts } from "./pages/Contacts";
import { Cart } from "./pages/Cart";

const AppRouter = function () {
    return (
        <Switch>
            <Route path="/" exact component={Home} />
            <Route path="/catalog" exact component={Catalog} />
            <Route path="/catalog/:id" component={Product} />
            <Route path="/delivery" component={Delivery} />
            <Route path="/contacts" component={Contacts} />
            <Route path="/cart" component={Cart} />
        </Switch>
    );
}

export default AppRouter;