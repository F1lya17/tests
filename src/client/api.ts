import axios, { AxiosResponse } from 'axios';
import { CartState, CheckoutFormData, CheckoutResponse, Product, ProductShortInfo } from '../common/types';

export abstract class AbstractApi {
    constructor(protected readonly basename: string) {

    }

    abstract getProducts(): Promise<AxiosResponse<ProductShortInfo[]>>

    abstract getProductById(id: number): Promise<AxiosResponse<Product | null>>

    abstract checkout(form: CheckoutFormData, cart: CartState): Promise<AxiosResponse<CheckoutResponse>>
}

export class MockApi extends AbstractApi {
    constructor(protected readonly basename: string, private readonly products: ProductShortInfo[]) {
        super(basename);
        products = products;
    }

    getProducts(): Promise<AxiosResponse<ProductShortInfo[], any>> {
        return new Promise((resolve): void => {
            resolve({
                status: 200,
                statusText: "good",
                data: this.products,
                headers: {},
                config: {},
            })
        });
    }

    getProductById(id: number): Promise<AxiosResponse<Product, any>> {
        return new Promise((resolve) => {
        });
    }

    checkout(form: CheckoutFormData, cart: CartState): Promise<AxiosResponse<CheckoutResponse, any>> {
        return new Promise((resolve) => {
        });
    }
}

export class ExampleApi extends AbstractApi {
    constructor(protected readonly basename: string) {
        super(basename);
    }
    async getProducts() {
        let res = await axios.get<ProductShortInfo[]>(`${this.basename}/api/products`);
        return res;
    }

    async getProductById(id: number) {
        return await axios.get<Product>(`${this.basename}/api/products/${id}`);
    }

    async checkout(form: CheckoutFormData, cart: CartState) {
        return await axios.post<CheckoutResponse>(`${this.basename}/api/checkout`, { form, cart });
    }
}

export const LOCAL_STORAGE_CART_KEY = 'example-store-cart';

export abstract class CartAbstractApi {
    constructor(protected readonly cart: CartState) {
        this.cart = cart
    }
    abstract getState(): CartState

    abstract setState(cart: CartState): void
}

export class CartMockApi extends CartAbstractApi {
    getState() {
        return this.cart;
    }
    setState(cart: CartState) { }
}

export class CartApi extends CartAbstractApi {
    getState(): CartState {
        try {
            const json = localStorage.getItem(LOCAL_STORAGE_CART_KEY);
            return JSON.parse(json) as CartState || {};
        } catch {
            return {};
        }
    }

    setState(cart: CartState) {
        localStorage.setItem(LOCAL_STORAGE_CART_KEY, JSON.stringify(cart));
    }
}