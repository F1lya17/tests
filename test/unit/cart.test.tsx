import React from 'react';
import { Provider } from 'react-redux';

import { ExampleApi, CartApi, MockApi, CartMockApi } from '../../src/client/api';
import { initStore } from '../../src/client/store';

import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom'
import { BrowserRouter } from 'react-router-dom';
import { Cart } from '../../src/client/pages/Cart';
import { Application } from '../../src/client/Application';

const basename = '/';

describe('Корзина', () => {
    it('в шапке рядом со ссылкой на корзину должно отображаться количество не повторяющихся товаров в ней', async () => {
        const api = new MockApi(basename, [{ id: 777, name: "Anton", price: 100 }]);
        const cart = new CartMockApi({ 777: { name: "Anton", price: 100, count: 2 } });
        const store = initStore(api, cart);
        render(
            <BrowserRouter basename=''>
                <Provider store={store}>
                    <Application />
                </Provider>
            </BrowserRouter>
        );
        expect(screen.getByTestId('cart')).toContainHTML('Cart (1)')
    })

    it('в корзине должна отображаться таблица с добавленными в нее товарами', async () => {
        const api = new MockApi(basename, [{ id: 777, name: "Anton", price: 100 }]);
        const cart = new CartMockApi({ 777: { name: "Anton", price: 100, count: 2 } });
        const store = initStore(api, cart);
        render(
            <BrowserRouter basename=''>
                <Provider store={store}>
                    <Cart />
                </Provider>
            </BrowserRouter>
        );
        await waitFor(() => {
            expect(screen.getByTestId('table')).toBeInTheDocument();
        })
    })

    it('для каждого товара должны отображаться название, цена, количество , стоимость, а также должна отображаться общая сумма заказа', async () => {
        const api = new MockApi(basename, [{ id: 777, name: "Anton", price: 100 }]);
        const cart = new CartMockApi({ 777: { name: "Anton", price: 100, count: 2 } });
        const store = initStore(api, cart);
        render(
            <BrowserRouter basename=''>
                <Provider store={store}>
                    <Cart />
                </Provider>
            </BrowserRouter>
        );
        await waitFor(() => {
            expect(screen.getAllByRole('product')).toMatchSnapshot();
        })
    })

    it('в корзине должна быть кнопка "очистить корзину", по нажатию на которую все товары должны удаляться', async () => {
        const api = new MockApi(basename, [{ id: 777, name: "Anton", price: 100 }]);
        const cart = new CartMockApi({ 777: { name: "Anton", price: 100, count: 2 } });
        const store = initStore(api, cart);
        render(
            <BrowserRouter basename=''>
                <Provider store={store}>
                    <Cart />
                </Provider>
            </BrowserRouter>
        );
        const btn = screen.getByTestId('clear');
        expect(btn).toBeInTheDocument();
        fireEvent.click(btn);
        await waitFor(() => {
            expect(screen.queryByRole('product')).toBeNull();
        })
    })

    it('в корзине должна быть кнопка "очистить корзину", по нажатию на которую все товары должны удаляться', async () => {
        const api = new MockApi(basename, [{ id: 777, name: "Anton", price: 100 }]);
        const cart = new CartMockApi({});
        const store = initStore(api, cart);
        render(
            <BrowserRouter basename=''>
                <Provider store={store}>
                    <Cart />
                </Provider>
            </BrowserRouter>
        );
        await waitFor(() => {
            expect(screen.getByTestId('empty')).toBeInTheDocument();
        })
    })
})