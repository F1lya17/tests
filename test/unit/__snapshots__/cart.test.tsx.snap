// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Корзина для каждого товара должны отображаться название, цена, количество , стоимость, а также должна отображаться общая сумма заказа 1`] = `
[
  <tr
    data-testid="777"
    role="product"
  >
    <th
      class="Cart-Index"
      scope="row"
    >
      1
    </th>
    <td
      class="Cart-Name"
    >
      Anton
    </td>
    <td
      class="Cart-Price"
    >
      $
      100
    </td>
    <td
      class="Cart-Count"
    >
      2
    </td>
    <td
      class="Cart-Total"
    >
      $
      200
    </td>
  </tr>,
]
`;
